### ISSUE TITLE
Make PET play nice with the rest of the code

### ISSUE THREAD

**Issue by tbohn** [UW-Hydro/VIC#276]
Currently, vic computes the 6 types of potential evapotranspiration (PET) by creating dummy veg tiles.  These dummy tiles cause code maintenance headaches and should be removed.  PET needs to be computed in a much less invasive manner.

**Comment by jhamman** [comment]
I've thought about this and I think there are probably things we can do here:

1.  Remove the PET calculation all together.  Yes this is a backward incompatable change but we can do that for VIC 5.  Calculating the PET as a post (or pre) process of VIC is probably 20 lines of Python code.  Since the value will never be used in the VIC model evolution, this seems to be the cleanest way to go.  
2.  Calculate PET for each subgrid tile, then use `put_data` (or whatever that ends up being) to average the tiles.  From my discussion with @tbohn - it seems like one of the reasons PET was added as it was, was to allow the PET calculation to use the VIC calculated met forcings and resistances.  So, in my view, to do this "right" would require pushing the PET code to the subgrid level.

All that said, @bartnijssen outlined these idioms for how we should approach the problem if we decide to keep the PET calcs in VIC:

1. unambiguous
2. general enough to be useful to a broad audience and not hard-wired for a specific project
3. easily maintainable (which it currently is not)
4. no memory allocated if the feature is not used (this is an issue for the image mode)
5. has a series of tests associated with it
6. documented

So @tbohn, can meet all these while keeping PET?

**Comment by tbohn** [comment]
To address ambiguity and generality, I propose to reduce the number of PET types to just what's currently called "PET_NATVEG", which is the type that people most commonly think of when looking at PET.  Several other models (NOAH off the top of my head) that I'm familiar with also provide this type of PET as an output variable so I don't think this would be unreasonable or unusual to include.  I added the other types in response to specific needs/requests but they were really one-offs and should probably be eliminated.

Treating PET the same as ET (as opposed to a dummy tile) should address maintainability.

I can make its allocation contingent on its being requested for output; that should address memory.

And I'll come up with tests and documentation - probably will need to discuss these in more detail offline.

**Comment by bartnijssen** [comment]
As long as it plays nice with others (especially no memory impact or requirements if the feature is not used) ...

**Comment by tbohn** [comment]
I'm thinking of addressing optional allocation this way: pot_evap will continue be part of the cell_data structure.  It will change from an array of fixed length to a *double.  If OUT_PET is selected for output, *pot_evap will be allocated to a single element of type double.  Else *pot_evap will not be allocated.

How does that sound?

**Comment by tbohn** [comment]
Regarding the calculation of pot_evap: I'm assuming we'll want to use the penman-monteith equation with canopy resistance computed using:
* rs = rsmin (rather than 0)
* LAI = current LAI
* all other resistance factors = 1.0.
Any objections?

**Comment by tbohn** [comment]
This may be a separate issue, but looking through the code, I see that for canopy evap, vic uses a canopy resistance (rc) value of 0; while for transpiration, vic uses a canopy resistance value of rc = ( rsmin / LAI )  (which is then adjusted for limitations of vpd, soil moisture, temperature and radiation).

I think this might explain why actual ET sometimes goes higher than PET, since PET has always been computed more like transpiration with a non-zero rsmin.  Shouldn't canopy evap use ( rsmin / LAI ) like transpiration?

I'll implement PET as non-limited transpiration, using rc = rsmin / LAI.  But keep in mind that actual ET might exceed PET when the canopy is wet if we don't change the canopy evap formulation.

**Comment by tbohn** [comment]
Question: to be consistent with the possibility of partial veg cover fraction < 1.0, should I implement PET as a weighted average of vegcover * (PET from veg) + (1-vegcover) * (PET from soil) ?

If so, then I might want to move the PET calculation to be right next to the transpiration and soil evap calculations in func_surf_energy_bal() to take advantage of the logic doing the weighted averaging.

**Comment by tbohn** [comment]
On second thought, that would be computationally expensive as the potential evap calculation would then be executed during every step of the surface balance iteration.  I'll keep the PET calculation where it is.

**Comment by tbohn** [comment]
@jhamman and @bartnijssen I am almost done with the coding and testing of this change.  One issue I need your feedback on is which type of PET to compute.

I'm trying to decide between:
1. PET = the transpiration that would occur from the land cover as it currently is (with its current architectural resistance and LAI), but in the absence of limitations from soil moisture, vapor pressure deficit, or sub-optimal temperature and insolation (this would be the full Penman-Monteith equation but with canopy resistance set to rsmin/LAI)
2. PET = same as option 1 but with canopy resistance set to 0 (this would be the Penman equation)

One nice thing about option 2 is that it has less dependence on vegetation parameters - no need to know LAI, stomatal, or architectural resistances (the net radiation and the aerodynamic resistance would still be affected by land cover, though).

Another nice thing about option 2 is that it's what is used to compute canopy evaporation in VIC.  If we use option 1 for PET, then it's possible for actual ET to exceed PET when the canopy is wet.

(I'm not sure I agree with VIC's using option 2 for canopy evap, since I'd think that architectural resistance and LAI would matter.)

The reference crop PETs that I implemented previously were presumably motivated by a desire for standard canopy and aerodynamic resistances, to get the best of both worlds - independent of actual land cover but more accurate than assuming zero resistance.  But then this just pushes the need for standardization to the choice of reference crop.

@bartnijssen from your experience, which option is most "standard" or most likely to be desired by the user?  What would you prefer?

**Comment by jhamman** [comment]
closed via #352
